data/transactions/application.go

```diff
@@ -0,0 +1,776 @@
+// call clone on each AppLocalState -- callers must do that for any values
+// where they intend to modify a contained reference type e.g. KeyValue.
+func cloneAppLocalStates(m map[basics.AppIndex]basics.AppLocalState) map[basics.AppIndex]basics.AppLocalState {
+	res := make(map[basics.AppIndex]basics.AppLocalState, len(m))
+	for k, v := range m {
+		res[k] = v
+	}
+	return res
+}
+
+// Allocate the map of AppParams if it is nil, and return a copy. We do *not*
+// call clone on each AppParams -- callers must do that for any values where
+// they intend to modify a contained reference type e.g. the GlobalState.
+func cloneAppParams(m map[basics.AppIndex]basics.AppParams) map[basics.AppIndex]basics.AppParams {
+	res := make(map[basics.AppIndex]basics.AppParams, len(m))
+	for k, v := range m {
+		res[k] = v
+	}
+	return res
+}
+
+// getAppParams fetches the creator address and AppParams for the app index,
+// if they exist. It does *not* clone the AppParams, so the returned params
+// must not be modified directly.
+func getAppParams(balances Balances, aidx basics.AppIndex) (params basics.AppParams, creator basics.Address, exists bool, err error) {
+	creator, exists, err = balances.GetAppCreator(aidx)
+	if err != nil {
+		return
+	}
+
+	// App doesn't exist. Not an error, but return straight away
+	if !exists {
+		return
+	}
+
+	record, err := balances.Get(creator, false)
+	if err != nil {
+		return
+	}
+
+	params, ok := record.AppParams[aidx]
+	if !ok {
+		// This should never happen. If app exists then we should have
+		// found the creator successfully.
+		err = fmt.Errorf("app %d not found in account %s", aidx, creator.String())
+		return
+	}
+
+	return
+}
+
+func applyDelta(kv basics.TealKeyValue, stateDelta basics.StateDelta) error {
+	if kv == nil {
+		return fmt.Errorf("cannot apply delta to nil TealKeyValue")
+	}
+
+	// Because the keys of stateDelta each correspond to one existing/new
+	// key in the key/value store, there can be at most one delta per key.
+	// Therefore the order that the deltas are applied does not matter.
+	for key, valueDelta := range stateDelta {
+		switch valueDelta.Action {
+		case basics.SetUintAction:
+			kv[key] = basics.TealValue{
+				Type: basics.TealUintType,
+				Uint: valueDelta.Uint,
+			}
+		case basics.SetBytesAction:
+			kv[key] = basics.TealValue{
+				Type:  basics.TealBytesType,
+				Bytes: valueDelta.Bytes,
+			}
+		case basics.DeleteAction:
+			delete(kv, key)
+		default:
+			return fmt.Errorf("unknown delta action %d", valueDelta.Action)
+		}
+	}
+	return nil
+}
+
+// applyStateDeltas applies a basics.EvalDelta to the app's global key/value
+// store as well as a set of local key/value stores. If this function returns
+// an error, the transaction must not be committed.
+func (ac *ApplicationCallTxnFields) applyStateDeltas(evalDelta basics.EvalDelta, params basics.AppParams, creator, sender basics.Address, balances Balances, appIdx basics.AppIndex, errIfNotApplied bool) error {
```

Comment:
Nit: It seems slightly more intuitive to me to call this `applyEvalDelta` and to call `applyDelta` instead `applyStateDelta`.

Reply:
Also nit: it would be good to document `errIfNotApplied`.

Reply:
Alternatively, instead of having `errIfNotApplied`, you could return different error types depending on whether you want your caller to filter out the error you returned and let the caller decide whether to drop the error.

Reply:
renamed methods and added docs for `errIfNotApplied`